hierarqcal/core.py

```diff
@@ -636,23 +650,188 @@ def __eq__(self, other):
+                    ]
+        else:
+            pivot_pattern_fn = pivot_pattern
+        return pivot_pattern_fn
+
+    def __call__(self, Qc_l, *args, **kwargs):
+        """
+        Call the motif, this generates the edges and qubits of the motif (directed graph) based on it's available qubits.
+        Each time a motif in the stack changes, a loop runs through the stack from the beginning and calls each motif to update the graph (the available qubits, the edges etc).
+
+        Args:
+            Qc_l (list): List of available qubits.
+            *args: Variable length argument list.
+            **kwargs: Arbitrary keyword arguments, such as:
+                * mapping (tuple(function, int)): TODO update this docstring, it's not a tuple anymore.
+                    Function mapping is specified as a tuple, where the first argument is a function and the second is the number of symbols it uses. A symbol here refers to an variational paramater for a quantum circuit, i.e. crz(theta, q0, q1) <- theta is a symbol for the gate.
+
+        Returns:
+            Qconv: Returns the updated version of itself, with correct nodes and edges.
+        """
+
+        self.pivot_pattern_fn = self.get_pivot_pattern_fn(self.pattern, Qc_l)
+        pivot_q = self.pivot_pattern_fn(Qc_l)
+        remaining_q = [q for q in Qc_l if q not in pivot_q]
+
+        if self.arity > 1:
+            nq_available = len(remaining_q)
+            if nq_available > 0:
+                if self.stride % nq_available == 0:
+                    # TODO make this clear in documentation
+                    # warnings.warn(
+                    #     f"Stride and number of available qubits can't be the same, received:\nstride: {self.stride}\n available qubits:{nq_available}. Defaulting to stride of 1"
+                    # )
+                    self.stride = 1
+                if self.boundary == "open":
+                    mod_nq = lambda x: x % nq_available
+                    Ec_l = [
+                        tuple(
+                            (
+                                remaining_q[i + j * self.stride]
+                                for j in range(self.arity-1)
+                                if i + j * self.stride < nq_available
+                            )
+                        )
+                        for i in range(self.offset, nq_available, self.step)
+                    ]
+                    
+
+                else:
+                    mod_nq = lambda x: x % nq_available
+                    Ec_l = [
+                        tuple(
+                            (Qc_l[mod_nq(i + j * self.stride)] for j in range(self.arity-1))
+                        )
+                        for i in range(self.offset, nq_available, self.step)
+                    ]
+
+                # Add the pivot qubit to each edge
+                Ec_l = Ec_l[: len(Ec_l) - len(Ec_l) % len(pivot_q)]
+                Ec_l = [
+                    tuple(edge+(p,))
```

Comment:
I think we should enable control over where (p,) gets wedged into the edge. I was thinking along the lines of a local pattern where "010" indicates the pivot goes in the middle for arity 3. So local parity must be size of arity, and then the edges from the two sets are merged based on the local pattern. 

This will help us to later enable multi "pivots" so for example if a arity 4 unitary is sent with local_pattern="0011", we obtain two partitions, both gets connected within themselves by a cycle pattern and merged based on the local pattern i.e. if "0011" -> "o"=[(1,2),(3,4)]   "p"=[(5,6),(7,8)] -> [(1,2,5,6),(3,4,7,8)] .
If local pattern is "1001" then [(5,1,2,6),(7,3,4,8)]. 

But I haven't fully fledged this out yet, specifically each partition needs a cycle to specify how the are connected within, then we need another cycle to specify how they are connected between and then finally the pattern helps merge them. There's a simpler structure here, but haven't found it yet. 

For now, just focus on the pivots being of size one, i.e. only local patterns of the form "01", "001", "00100" (only one 1). i.e. the simplest solution where we can still just use `p` in this list comprehension and not worry about it being a tuple.

Reply:
I think it's working now. Have tested on 3-qubit unitary "cnot()^01;cnot()^21", see the last tests in [playtest_qiskit_qpivot](https://github.com/AmyRouillard/hierarqcal/blob/feature/qft/tests/playtest_qiskit_qpivot.py).